### ISSUE TITLE
give every widget clickRoutine

### ISSUE THREAD

**Issue by robartsd** [ArnoldSmith86/virtualtabletop#185]
I have the radical idea of moving the clickRoutine code to widget.js. Most widgets would have a default empty click routine. Cards and Basic Widgets would have default click routines to flip themselves. We'd keep button widget type mostly as a set of default properties, but this would allow any widget type to become a button. Our new dice would be Basic Widgets with special click routines instead of needing a separate button to make things work. This also might be helpful for refactoring automation code to use in onEnter and onLeave routines.

**Comment by pf-lkobbe** [comment]
This idea has been discussed on several occasions in our meetings and our conclusion was this would indeed mean a too "radical" change for VTT. The problematic aspect with this approach is that when you play a new game created by someone else, you would no longer know what clicking any widget might do (or not), possibly leading to a lot of confusion (and people randomly clicking widgets to see what the result is).

For the new dice, I would suggest that we implement them as Cards, and once we've implemented the Deck Editor, these Dice would just become one of the pre-defined decks, so it'd be easy for anyone to include standard dice in their game.

**Comment by larkob** [comment]
This idea has been discussed on several occasions in our meetings and our conclusion was this would indeed mean a too "radical" change for VTT. The problematic aspect with this approach is that when you play a new game created by someone else, you would no longer know what clicking any widget might do (or not), possibly leading to a lot of confusion (and people randomly clicking widgets to see what the result is).

For the new dice, I would suggest that we implement them as Cards, and once we've implemented the Deck Editor, these Dice would just become one of the pre-defined decks, so it'd be easy for anyone to include standard dice in their game.

**Comment by 96LawDawg** [comment]
Dice as cards may or may not work because of stacking.  I'm not talking about the bigger problem with piles, but just the fact that cards tend to want to stick together and if the dice are their own widgets, then there is no pile issue.

**Comment by robartsd** [comment]
With transparent buttons, the game designer can already (visually) make it so that you don't know what clicking will do. Of course even an relatively advanced user looking at the DOM might have difficulty finding what was going on in the javascript.

**Comment by RaphaelAlvez** [comment]
As I said in discord, even though this may be a good fit it is not as simple as just "add click routine

**Comment by robartsd** [comment]
> This idea has been discussed on several occasions in our meetings and our conclusion was this would indeed mean a too "radical" change for VTT. The problematic aspect with this approach is that when you play a new game created by someone else, you would no longer know what clicking any widget might do (or not), possibly leading to a lot of confusion (and people randomly clicking widgets to see what the result is).

I can understand this reasoning. If this is the reasoning we're following, we should disable all game styles in edit mode. A designer could still make invisible buttons that make any widget appear to have extra functionality, but a quick trip to edit mode would fairly easily show any user information about what is going on.

If this were my project I would choose to implement this, but that's just my opinion and not a reason I'd leave the project over.

**Comment by robartsd** [comment]
> For the new dice, I would suggest that we implement them as Cards, and once we've implemented the Deck Editor, these Dice would just become one of the pre-defined decks, so it'd be easy for anyone to include standard dice in their game.

This is not a solution because they need a button to apply the movement cue that they have changed when the result is the same (see #151 ). The problem isn't dice image assets it is dice behavior. We can extend widgets to add the needed behavior, create a new widget with the behavior, or provide compound widgets with buttons that define the behavior.

**Comment by 96LawDawg** [comment]
For what it's worth, I agree with robartsd about this proposal and philosophy.  I have already made and implemented invisible buttons in library games so that I could implement a click functionality on a better looking image than I can get with the default button toolset.  I've actually been slightly disappointed that clickRoutine is not available in basic widgets since first coming to the project.  But that's just my observation.

Let game designers do what they want.  We've said it before...if they build unusable crappy games, no one will play them and we don't have to put them in the library.

**Comment by robartsd** [comment]
For how I want to use dice, this would be easiest to build games, but #184 would be functional enough. A new Dice widget that flips to a random face and rotates a random when clicked would work too.

@larkob I'm not clear on why you prefer Card vs. Basic Widget, in my mind Card is a member of a deck - very useful when making a bunch of similar widgets. A deck of dice would be good for games that use a lot of dice, dice from Basic Widgets would better serve games that only need a few.

**Comment by shaunhillier** [comment]
I added a lot of invisible buttons to splendid and the annoying thing about is you lose the hover over enlargement, if cards had their own click routine I think it would make for a better User experience.
I don't know if it is possible without a style sheet, but could we do something clever where if a widget has a non default click routine the mouse changes, as a sign that clicking on this thing will do something?

**Comment by larkob** [comment]
This is a difficult topic for me, because I was the one strongly advocating for clickRoutines in all widget types in the early stages of VTT. So, I fully understand anyone's current disappointment/disagreement. 

I also agree with invisible buttons feeling very much like a hack for the real thing (clickRoutines). Nevertheless, I believe that we should carefully consider alternative approaches before going for universal clickRoutines. I actually have a proposal of how to accomplish what you'd want for dice without the need for clickRoutines, but I want to discuss it in its own issue.

**Comment by larkob** [comment]
> @larkob I'm not clear on why you prefer Card vs. Basic Widget, in my mind Card is a member of a deck - very useful when making a bunch of similar widgets. A deck of dice would be good for games that use a lot of dice, dice from Basic Widgets would better serve games that only need a few.

My reasoning is this: Once the Deck Editor is ready, it'll be a piece of cake to create any type and amount of dice for any game. What's more, it would also facilitate sharing dice designs. I'm happy to present this idea with some visual mockups, but that discussion should probably take place in its own issue.
